api/src/main/java/org/apache/iceberg/IncrementalTableScan.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+package org.apache.iceberg;
+
+/**
+ * API for configuring an incremental table scan
+ */
+public interface IncrementalTableScan extends Scan<IncrementalTableScan> {
+
+  /**
+   * Optional. if not set, null value will be used for the start snapshot id.
+   * That would include the oldest ancestor of the {@link IncrementalTableScan#toSnapshotId(long)},
+   * as its parent snapshot id is null which matches the null start snapshot id
+   *
+   * @param fromSnapshotId the start snapshot id (exclusive)
+   * @return an incremental table scan from {@code fromSnapshotId} exclusive
+   */
+  IncrementalTableScan fromSnapshotId(long fromSnapshotId);
+
+  /**
+   * Optional. if not set, current table snapshot id is used as the end snapshot id
+   *
+   * @param toSnapshotId the end snapshot id (inclusive)
+   * @return an incremental table scan up to {@code toSnapshotId} inclusive
+   */
+  IncrementalTableScan toSnapshotId(long toSnapshotId);
+
+  /**
+   * Only interested in snapshots with append operation
+   */
+  IncrementalTableScan appendsOnly();
+
+  /**
+   * Ignore snapshots with overwrite operation.
+   *
+   * Default behavior for incremental scan fails if there are overwrite operations in the incremental snapshot range
+   */
+  IncrementalTableScan ignoreOverwrites();
```

Comment:
For V2 tables, incremental scan is potentially interested in append, overwrite, delete operations. append is always included. I guess we need to discuss what is the natural API to control overwrite and delete.

this API of `ignoreOverwrites` might be too restrictive. then we also might need to expose an `includeOverwrites`. I am wondering we should api like
```
include(DataOperations... operations) // default is append only. for CDC read, overwrite and delete can be added
fail(DataOperations... operations) // default is fail nothing
```

Reply:
or maybe we should just need the `appendsOnly` API above. otherwise, overwrite and delete snapshots are also included automatically for V2 tables.

Reply:
> or maybe we should just need the `appendsOnly` API above. otherwise, overwrite and delete snapshots are also included automatically for V2 tables.

 I agree. Because i think for V2 table include overwrite and delete snapshots should be a default behavior.

Reply:
For the logical completeness of filtering snapshot type, should we have the following 3 methods?
1. ignoreAppend
2. ignoreOverwrite
3. ignoreDelete

And the `appendsOnly` is basically a combination of `ignoreOverwrite` and `ignoreDelete`

Reply:
agree the 3 ignore methods are complete and flexible? wondering if we need the combinations or we only need two modes (appendsOnly and append + overwrite + delete). E.g., does `append + overwrite` or `overwrite + delete` ever make sense?

Do we need fail APIs (like `failOverwrite`)? My current take is no. Let's just skip the snapshots (not interested).

Reply:
Yes, it is flexible. `append + overwrite` makes sense for user who only want to get inserted rows with some additional filtering. `overwrite + delete` makes sense for getting only deleted rows.
I'm not aware of a use case with `failOverwrite`. We may skip it now.

Reply:
How about dataBetween(Long startSnapshotId, long endSnapshot, List<RowKind> rowKinds)?

Reply:
@chenjunjiedada `RowKind` is a Flink API. 

Based on the conversation, it seems most people prefer the fluent style API for the scan builder, like `fromSnapshotId(long fromSnapshotId)`

Reply:
I meant we could borrow the `RowKind` definition of the produced data. Like what @flyrain mentioned before. We could use `+I` to target the `append` + `overwrite` with some filter, use `-D, -U` to target `delete` and some data in `overwrite`.